model/riscv_insts_base.sail

```diff
@@ -321,28 +321,86 @@ function check_misaligned(vaddr : xlenbits, width : word_width) -> bool =
          WORD   => vaddr[0] == bitone | vaddr[1] == bitone,
          DOUBLE => vaddr[0] == bitone | vaddr[1] == bitone | vaddr[2] == bitone
        }
+function transform_VA (effective_address : xlenbits,PMLEN :int) -> xlenbits = {
+  assert(PMLEN >= 0, "PMLEN out of range");
+  assert(PMLEN <= 63, "PMLEN out of range");
+  let XLEN = sizeof(xlen);
+  let i = XLEN -(PMLEN +1);
+  let effective_bit = effective_address[i..i];
+  let masked_bits = replicate_bits(effective_bit, PMLEN);
+  let unmasked_bits = effective_address[i..0];
+  masked_bits @ unmasked_bits
+}
+
+function transform_PA (effective_address : xlenbits,PMLEN :int) -> xlenbits = {
+  assert(PMLEN >= 0, "PMLEN out of range");
+  assert(PMLEN <= 63, "PMLEN out of range");
+  let XLEN = sizeof(xlen);
+  let i = XLEN -(PMLEN + 1);
+  let masked_bits = replicate_bits(0b0, PMLEN);
+  let unmasked_bits = effective_address[i..0];
+  masked_bits @ unmasked_bits
+}
+
+function transform_effective_address(vaddr : xlenbits, pmm : bool) -> xlenbits = {
+ if pmm then {
+    let pmm_bits : bits(2) = get_pmm();
+    let mode : SATPMode = translationMode(cur_privilege);
+    match (mode) {
+      (Sbare) => {
+        match (pmm_bits){
+          (0b10) => transform_PA (vaddr,7),
+          (0b11) => transform_PA (vaddr,16),
+        }
+      },
+      (Sv39) => {
+        match (pmm_bits){
+          (0b10) => transform_VA (vaddr,7),
+          (0b11) => transform_VA (vaddr,16),          
+        }
+      },      
+      (Sv48) => {
+        match (pmm_bits){
+          (0b10) => transform_VA (vaddr,7),
+          (0b11) => transform_VA (vaddr,16),          
+        }
+      },
+      (Sv57) => {
```

Comment:
Did you really intend all of these branches to be identical? That doesn't seem right.

Reply:
Basically here i intended to write a code in a way that if we are in Bare mode then then mask the Physical Address  and if we are in the any Translation Scheam mode then Mask the Virtaul Adrees  base on the value of PMM bits.BUT when i wrote a code in this way : `    match mode {
      Sbare => {
        match pmm_bits {
          0b10 => return transform_PA(vaddr, 7),
          0b11 => return transform_PA(vaddr, 16),
          _    => return vaddr // Default case
        }
      },
      Sv39 | Sv48 | Sv57 => {
        match pmm_bits {
          0b10 => return transform_VA(vaddr, 7),
          0b11 => return transform_VA(vaddr, 16),
          _    => return vaddr // Default case
        }
      },`

Reply:
Ah I missed the PA/VA difference. Yes I would do it either like this:

```
if mode == Sbare then {
...
} else {
...
}
```

Or better:

```
match pmm_bits {
  0b10 => if mode == Sbare then transform_PA(vaddr, 7) else transform_VA(vaddr, 7),
  0b11 => if mode == Sbare then transform_PA(vaddr, 16) else transform_VA(vaddr, 16),
  _ => vaddr,
}
```

Note you don't need the `return`.

You could also do this but I don't know if it's clearer. Debatable.

```
match pmm_bits {
  0b1 @ size_or_whatever : bits(1) => {
     let other_size_thing = if size_or_whatever == 0b0 then 7 else 16;
     if mode == Sbare then transform_PA(vaddr, other_size_thing ) else transform_VA(vaddr, other_size_thing )
  },
  _ => vaddr,
}
```

Reply:
Done